volume/volumefs/layout.go

```diff
@@ -0,0 +1,164 @@
+// for details on the number of blocks that can be tracked. e.g.
+//   Depth     N=128      N=256
+//     3         2Mi      16Mi
+//     4       256Mi    4096Mi
+//
+// There is a fixed number of blocks that a symmetric tree can track. To preserve the
+// ability to increase the number of blocks in the future while still remaining backward
+// compatible we add extra address space potential to top-level directory entries by adding
+// leading zeros to their name.  Then in the future we have the choice to add additional
+// trees of depth maxTreeDepth-1 or introduce indirect blocks for additional trees of
+// depth maxTreeDepth.  See the unit test for illustration.
+
+const (
+	maxDirEntries     = int64(256) // nolint:gomnd // must be a power of 2
+	maxTreeDepth      = 3
+	snapshotBlockSize = int64(1024 * 1024) // nolint:gomnd
+	extraTopHexFmtLen = 1
+	fileHexFmtLen     = 12 // counts up to 2^48
+	hexBits           = 4
+)
+
+type layoutProperties struct {
+	blockSzB int64
+	dirSz    int64
+	depth    int
+	// the rest are computed
+	dirSzL2         int
+	dirSzMask       int64
+	maxBlocks       int64
+	maxVolSzB       int64
+	dirHexFmtLen    int
+	dirTopHexFmtLen int
+	fileHexFmtLen   int
+}
+
+func (f *Filesystem) setDefaultLayoutProperties() {
+	f.layoutProperties.initLayoutProperties(snapshotBlockSize, maxDirEntries, maxTreeDepth)
+}
+
+func (l *layoutProperties) initLayoutProperties(snapshotBlockSize, maxDirEntries int64, maxTreeDepth int) {
+	l.blockSzB = snapshotBlockSize
+	l.dirSz = maxDirEntries
+	l.depth = maxTreeDepth
+	l.dirSzL2 = int(math.Log2(float64(maxDirEntries)))
+	l.dirSzMask = maxDirEntries - 1
+	l.maxBlocks = 1 << (maxTreeDepth * l.dirSzL2)
+	l.maxVolSzB = l.maxBlocks * l.blockSzB
+	l.dirHexFmtLen = l.dirSzL2/hexBits + (l.dirSzL2%4+3)/hexBits
+	l.dirTopHexFmtLen = l.dirHexFmtLen + extraTopHexFmtLen
+	l.fileHexFmtLen = fileHexFmtLen // fixed
+}
+
+// GetBlockSize returns the snapshot block size
+func (f *Filesystem) GetBlockSize() int64 {
+	return f.blockSzB
+}
+
+type parsedPath []string
+
+func (pp parsedPath) String() string {
+	return path.Join(pp...)
+}
+
+// addrToPath returns the parsed path for a block file
+func (f *Filesystem) addrToPath(blockAddr int64) (parsedPath, error) {
+	if blockAddr >= f.maxBlocks || blockAddr < 0 {
+		// @TODO: future special case blockAddr > f.maxBlocks with reserved hex digit in top dir
+		// Can choose to add additional trees of depth maxTreeDepth-1 or add indirect blocks for
+		// additional trees of depth maxTreeDepth.
+		return nil, ErrOutOfRange
+	}
+
+	pp := make(parsedPath, 0, maxTreeDepth)
+
+	for i, dirHexFmtLen := maxTreeDepth-1, f.dirTopHexFmtLen; i > 0; i, dirHexFmtLen = i-1, f.dirHexFmtLen {
+		idx := (blockAddr >> (f.dirSzL2 * i)) & f.dirSzMask
+		el := fmt.Sprintf("%0*x", dirHexFmtLen, idx)
+		pp = append(pp, el)
+	}
+
+	pp = append(pp, fmt.Sprintf("%0*x", f.fileHexFmtLen, blockAddr))
```

Comment:
After discussion with @julio-lopez we should switch to directory relative naming to minimize the total size of the directory in bytes.